optaplanner-core/src/test/java/org/optaplanner/core/config/solver/SolverConfigMultiThreadedTest.java

```diff
@@ -0,0 +1,119 @@
+package org.optaplanner.core.config.solver;
+
+import java.util.List;
+import java.util.stream.Collectors;
+import java.util.stream.IntStream;
+
+import org.junit.Test;
+import org.optaplanner.core.api.solver.Solver;
+import org.optaplanner.core.api.solver.SolverFactory;
+import org.optaplanner.core.config.solver.testutil.TestThreadFactory;
+import org.optaplanner.core.impl.testdata.domain.TestdataEntity;
+import org.optaplanner.core.impl.testdata.domain.TestdataSolution;
+import org.optaplanner.core.impl.testdata.domain.TestdataValue;
+import org.optaplanner.core.impl.testdata.util.PlannerTestUtils;
+
+import static org.assertj.core.api.Assertions.assertThat;
+import static org.assertj.core.api.Assertions.assertThatExceptionOfType;
+import static org.mockito.Mockito.spy;
+import static org.mockito.Mockito.when;
+
+public class SolverConfigMultiThreadedTest {
+
+    @Test
+    public void testMoveThreadCountAutoPositive() {
+        final int cpuCount = 16;
+        assertThat(mockSolverConfigForMoveThreadCountAuto(cpuCount).resolveMoveThreadCount()).isEqualTo(cpuCount - 2);
+    }
+
+    @Test
+    public void testMoveThreadCountAutoNegative() {
+        final int cpuCount = -2;
+        assertThat(mockSolverConfigForMoveThreadCountAuto(cpuCount).resolveMoveThreadCount()).isNull();
+    }
+
+    private SolverConfig mockSolverConfigForMoveThreadCountAuto(int mockCpuCount) {
+        SolverConfig solverConfig = spy(SolverConfig.class);
+        when(solverConfig.getAvailableProcessors()).thenReturn(mockCpuCount);
+        solverConfig.setMoveThreadCount(SolverConfig.MOVE_THREAD_COUNT_AUTO);
+        return solverConfig;
+    }
+
+    @Test
+    public void testMoveThreadCountFixedPositive() {
```

Comment:
test methods shouldn't start with "test" prefix, that's an artifact of JUnit 3 (and we're using JUnit 4, soon 5)

Reply:
I use name that can read as a sentence; that said, I may use passive form and omit the "test" prefix. As majority of existing tests already do so, will follow this convention.

Reply:
Not so recently I started to like the idea to make test methods as descriptive as possible even at the cost of breaking Java naming conventions. But that's OK, tests are not genuine Objects whose methods' purpose is sending messages.

For example:
```java
public void resolvedMoveThreadCountShouldEqualToConfig()
public void negativeMoveThreadCountShouldThrowException()
```
or even this more readable syntax:
```java
public void resolved_move_thread_count_should_equal_to_config()
public void negative_move_thread_count_should_throw_exception()
```

When reading the test method it is immediately clear what the test does and what's the intended behavior. It is also helpful when reading test report and looking at failures.

I understand that keeping the tests consistent might be a good reason for not adopting this style but I think it was worth mentioning.

Reply:
+1 for making method names as much descriptive as possible. There are several different conventions described, many of them working with "expected behaviour" and "state under test" [1]. However, if we adopt any of these conventions, we should rename the existing tests to ensure consistency.

[1] https://dzone.com/articles/7-popular-unit-test-naming